drizzle/resample.py

```diff
@@ -0,0 +1,404 @@
+        #     pmap_xmin = int(np.floor(np.min(pixmap[:, :, 0])))
+        #     pmap_xmax = int(np.ceil(np.max(pixmap[:, :, 0])))
+        #     pmap_ymin = int(np.floor(np.min(pixmap[:, :, 1])))
+        #     pmap_ymax = int(np.ceil(np.max(pixmap[:, :, 1])))
+        #     pixmap = pixmap.copy()
+        #     pixmap[:, :, 0] -= pmap_xmin
+        #     pixmap[:, :, 1] -= pmap_ymin
+        #     self.out_shape = (
+        #         pmap_xmax - pmap_xmin + 1,
+        #         pmap_ymax - pmap_ymin + 1
+        #     )
+
+        if xmin is None or xmin < 0:
+            xmin = 0
+
+        if ymin is None or ymin < 0:
+            ymin = 0
+
+        if xmax is None or xmax > self.out_shape[1] - 1:
+            xmax = self.out_shape[1] - 1
+
+        if ymax is None or ymax > self.out_shape[0] - 1:
+            ymax = self.out_shape[0] - 1
+
+        if weight_map is not None:
+            weight_map = np.asarray(weight_map, dtype=np.float32)
+        else:  # TODO: this should not be needed after C code modifications
+            weight_map = np.ones_like(data)
+
+        pixmap = np.asarray(pixmap, dtype=np.float64)
+
+        # TODO: probably tdriz should be modified to not return version.
+        #       we should not have git, Python, C, ... versions
+
+        # TODO: While drizzle code in cdrizzlebox.c supports weight_map=None,
+        #       cdrizzleapi.c does not. It should be modified to support this
+        #       for performance reasons.
+        if len(self.out_ctx.shape) == 2:
+            assert plane_no == 0
+            ctx_plane = self.out_ctx
+        else:
+            ctx_plane = self.out_ctx[plane_no]
+
+        _vers, nmiss, nskip = cdrizzle.tdriz(
+            data,
+            weight_map,
+            pixmap,
+            self.out_sci,
+            self.out_wht,
+            ctx_plane,
+            uniqid=id_in_plane + 1,
+            xmin=xmin,
+            xmax=xmax,
+            ymin=ymin,
+            ymax=ymax,
+            scale=scale, # scales image intensity. usually equal to pixel scale
+            pixfrac=pixfrac,
+            kernel=self.kernel,
+            in_units=in_units,
+            expscale=expscale,
+            wtscale=wht_scale,
+            fillstr=self.fillval
+        )
+        self._cversion = _vers  # TODO: probably not needed
+
+        return nmiss, nskip
+
+
+def blot_image(data, pixmap, pix_ratio, exptime, output_pixel_shape,
+               interp='poly5', sinscl=1.0):
+    """
+    Resample input image using interpolation to an output grid.
+    Typically, this is used to resample the resampled image that is the
+    result of multiple applications of ``Drizzle.add_image`` **if it is well
+    sampled** back to the pixel coordinate system of input (with distorted
+    WCS) images of ``Drizzle.add_image``.
+
+    Parameters
+    ----------
+
+    data : 2D array
+        Input numpy array of the source image in units of 'cps'.
+
+    pixmap : 3D array
```

Comment:
Is this the same array as used for drizzle or one that maps the other direction?

Reply:
it is "inverse" map: it maps from the `wcs_from` to `wcs_to`.